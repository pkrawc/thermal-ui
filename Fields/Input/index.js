import React, { Component, PropTypes } from 'react'
import styled from 'styled-components'
import { colors } from 'variables'
import shortid from 'shortid'

const InputGroup = styled.div`
  position: relative;
  overflow: hidden;
  input {
    padding: 2em 1em 1em;
    background: ${({dark}) => dark ? 'transparent' : '#F9F9F9'};
    border: ${({dark}) => dark ? '2px solid rgba(255,255,255,0.6)' : '2px solid rgba(0,0,0,0.12)'};
    border-radius: 2px;
    font-size: 1em;
    outline: none;
    width: 100%;
    transition: all 100ms ease-in-out;
    &:focus {
      border: 2px solid #137BB5;
      box-shadow: 0 3px 6px rgba(0,0,0,0.12);
      & + label { transform: translateY(-0.25em) scale(0.5); }
    }
    &.input-error {
      border: 2px solid ${colors.error};
      color: ${colors.error};
      & + label {
        color:${colors.error};
        opacity: 0.6;
      }
    }
    & + label {
      position: absolute;
      top: 0.75em;
      left: 1rem;
      font-size: 1.5em;
      transform-origin: top left;
      transform: ${({value, placeholder}) => placeholder ? 'translateY(-0.25em) scale(0.5)' : value.length > 0 ? 'translateY(-0.25em) scale(0.5)' : 'inital'};
      transition: all 100ms ease-in-out;
      color: ${({dark}) => dark ? 'rgba(255,255,255,0.6)' : 'rgba(0,0,0,0.4)'};
    }
  }
  span.error {
    color: ${colors.error};
    font-size: 0.75em;
  }
`

export default class TextInput extends Component {
  static propTypes = {
    defaultValue: PropTypes.string,
    error: PropTypes.bool,
    errorText: PropTypes.string,
    label: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    type: PropTypes.oneOf(['text', 'password', 'email', 'tel']).isRequired,
    icon: PropTypes.string,
  }
  static defaultProps = {
    label: 'Input',
    type: 'text',
    defaultValue: '',
    onChange: e => null
  }
  constructor(props) {
    super(props)
    this.id = shortid.generate()
    this.state = {value: this.props.defaultValue}
  }
  componentDidMount() {
    if (this.props.autofocus) this.input.focus()
  }
  render() {
    const {
      autofocus,
      defaultValue,
      error,
      errorText,
      label,
      onChange,
      placeholder,
      type,
      ...rest
    } = this.props
    return (
      <InputGroup {...rest} value={this.state.value} placeholder={placeholder}>
        <input
          className={ error ? 'input-error' : '' }
          defaultValue={defaultValue}
          id={this.id}
          onChange={e => {
            this.setState({value: e.target.value}),
            onChange(e.target.value)
          }}
          placeholder={placeholder}
          ref={ input => this.input = input }
          type={type}
          {...rest}
        />
        <label htmlFor={this.id}>{label}</label>
        {error ? <span className="error">Error: {errorText}</span> : null}
      </InputGroup>
    )
  }
}
