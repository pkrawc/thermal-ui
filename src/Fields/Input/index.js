import React, { PropTypes } from 'react'
import styled from 'styled-components'
import { colors } from 'variables'

const InputWrapper = styled.div`
  position: relative;
  padding: 1em 0 2em;
  input {
    padding: 0.5em 0;
    background: none;
    border: none;
    border-bottom: 2px solid ${({error, theme}) => error ? theme.colors ? theme.colors.error : colors.error : colors.disabled};
    color: ${({error, theme}) => error ? theme.colors ? theme.colors.error : colors.error : colors.dark};
    font-size: 1em;
    outline: none;
    width: 100%;
    transition: all 200ms ease-in-out;
    &:hover {
      border-bottom: 2px solid ${colors.darkSecondary};
    }
    &:focus {
      border-bottom: 2px solid ${colors.darkSecondary};
    }
    &::-webkit-input-placeholder {
       color: ${colors.disabled};
    }
  }
  input + label {
    position: absolute;
    top: 0;
    left: 0;
    color: ${({error, theme}) => error ? theme.colors ? theme.colors.error : colors.error : colors.darkSecondary};
    font-size: 0.75em;
    transition: all 200ms ease-in-out;
  }
  .error {
    position: absolute;
    bottom: 1em;
    right: 0;
    color: ${({theme}) => theme.colors ? theme.colors.error : colors.error };
    font-size: 0.75em;
    text-align: right;
    visibility: ${({error}) => error ? 'visible' : 'hidden'};
    opacity: ${({error}) => error ? '1' : '0'};
    transition: color 200ms ease-in-out, opacity 200ms ease-in-out;
  }
`

export const Input = ({
  id,
  type,
  value,
  placeholder,
  errorText,
  ...props
}) =>
  <InputWrapper {...props}>
    <input
      id={id}
      type={type}
      value={value}
      placeholder={placeholder}
    />
    { type === 'email' ? <span className="mdi mdi-email"></span> : null }
    <label htmlFor={props.id}>{props.label}</label>
    <span className="error">
      {errorText}
    </span>
  </InputWrapper>

Input.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  error: PropTypes.bool,
  errorText: PropTypes.string,
  placeholder: PropTypes.string,
}

Input.defaultProps = {
  type: 'text'
}
