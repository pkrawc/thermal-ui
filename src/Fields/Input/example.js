import React, { Component } from 'react'
import { Input } from '../Input'
import { Button } from '../../index'
import { colors } from 'variables'

class InputExample extends Component {
  constructor(props) {
    super(props)
    this.state = {
      error: false
    }
  }
  render() {
    return (
      <div>
        <Input
          id="example"
          type="text"
          label="Tenants Name"
          placeholder="Janet Wood"
          error={this.state.error}
          errorText="oh shit, you fucked up"
        />
        <Button onClick={() => {
          this.setState({error: !this.state.error})
        }}>toggle error</Button>
      </div>
    )
  }
}

export default {
  title: 'Input',
  example: InputExample
}
