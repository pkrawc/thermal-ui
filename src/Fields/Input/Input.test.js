import React from 'react'
import renderer from 'react-test-renderer'
import { Input } from '../Input'

describe('Input', () => {
  const props = {
    id: '1',
    label: 'input label',
    type: 'text'
  }
  const tree = renderer.create(<Input {...props} />).toJSON()

  it('renders without crashing', () => {
    expect(tree).toMatchSnapshot()
  })
})
