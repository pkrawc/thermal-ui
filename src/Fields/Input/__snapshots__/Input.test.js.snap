exports[`Input renders without crashing 1`] = `
<div
  className="bvTvgI"
  label="input label">
  <input
    id="1"
    placeholder={undefined}
    type="text"
    value={undefined} />
  <label
    htmlFor={undefined}>
    input label
  </label>
</div>
`;
